### ISSUE TITLE
InputFilter: allow_empty should check filtered input

### ISSUE THREAD

**Issue by hschletz** [zendframework/zf2#6668]
As a follow-up to my comments for #6133, I think the observed behavior for allow_empty is incorrect. The check is performed on unfiltered input, but the validator chain receives filtered input. This inconsistency can lead to erratic behavior if an attached filter alters input in a way that affects the definition of "emptiness".

For example, this happens with the StringTrim filter when raw input contains only whitespace. This is considered non-empty raw input, but the filter chain will produce an empty string. If allow_empty is TRUE, the validator chain still gets invoked which, depending on the particular validators, may cause validators to fail if these expect non-empty input. The validator chain should not be invoked in that case because empty input was explicitly declared valid.

If the StringTrim filter in the above example is followed by a Null filter, things can get really weird. Many validators (like StringLength) don't accept NULL input which is not a problem by itself, but causes the validator chain to fail when it should not have been invoked at all.

This may look like an edge case, but any filter that converts non-empty input to empty values or vice versa, or just produces a different kind of empty value (like empty string to NULL) can cause incorrect results.

Currently I have to work around the issue by reinventing the wheel: extend existing validators to re-check for empty values, or write a callback even for the most basic validations (see StringLength/Null example above).

All this hassle could be avoided by an additional emptiness check on the filter chain result - the same value that is passed to the validator chain. That would be the only behavior that works consistently and independently from the actual filter chain.

The continue_if_empty option has no effect on this behavior.

**Comment by mbn18** [comment]
I stumbled upon a strange behaviour, that even though the flag 'allow_empty" was set true the EmptyValidator was injected.

On https://github.com/zendframework/zf2/blob/master/library/Zend/InputFilter/Input.php#L361 it does the following check:
```
if ((!$this->isRequired() && $this->allowEmpty()) || $this->notEmptyValidator) {
    return;
}
```
In that session both required and allow_empty are set to true and it seems that it should return if $this->allowEmpty is true.

More strange, When it was called from the controller it worked as expected:
```
$inputFilter = new \Zend\InputFilter\InputFilter();
$inputFilter->add([
    'name'          => 'relatedPredKeys',
    'required'      => true,
    'allow_empty'   => true
]);
```
But from extended InputFilter class it faild.

Is that related?

**Comment by hschletz** [comment]
Possibly, but difficult to tell without more information on the failing case.

Could you test your code against PR #7056?

**Comment by hschletz** [comment]
Fixed with 2f271790ec62b02c3f2b892986ea05f201ef67e0.
